platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/RootTestEventRecorder.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.internal.tasks.testing.junit.result.TestReportDataCollector;
+import org.gradle.api.internal.tasks.testing.results.TestListenerAdapter;
+import org.gradle.api.internal.tasks.testing.results.TestListenerInternal;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Records test events for a single root test execution to a binary results directory.
+ */
+public class RootTestEventRecorder implements TestListenerInternal, Closeable {
```

Comment:
This will probably change anyways once Octavia's changes land, but I was thinking this would be slightly different.

Instead of a RootTestEventRecorder that's also a TestListenerInternal, we'd register the TestListenerAdapter below directly.

Then instead of passing the RootTestEventRecorder to the RootTestEventReporter, we'd pass the TestReportDataCollector and resultsDirectory explicitly.  This gets rid of the RootTestEventRecorder entirely. 

The reason being is that the delegation here is kind of hokey and the RootTestEventReporter is already deeply involved in managing the lifecycle of what's going on inside here.